### ISSUE TITLE
How to customise feature selection? Add new filters, use aggregate values from bootstrapping

### ISSUE THREAD

**Issue by andrewjohnlowe** [mlr-org/mlr#878]
Hello! I have two questions:

1. Is there a way to define my own filter-based feature selection? I'd like to use the filter methods in CORElearn. Is this possible? If so, how can I define and plug-in my own filter, or one from another package?

2. From past experience, I know that several of the filter-based measures that I'm interested in produce unstable results on my data. I'd like to bootstrap these measures and use an aggregate value -- mean or median -- instead. For example, info gain on my data seems to volatile, but the mean or median of a few tens of values evaluated on bootstrapped resamplings is stable. Ideally, I'd like to do bootstrapped filter-based feature selection (perhaps with a filter from another package; see Q1 above) as a step before learning, but fused with a learner, as part of nested cross-validation. I'm wondering if there is something like:
`lrn = makeFilterWrapper(learner = "classif.fnn", fw.method = "information.gain", fw.abs = 2, fw.boot = 100)`
for example. Is this possible? If so, how? (Would be nice to also access the SD or IQR of the measures too.)

Thanks!

**Comment by schiffner** [comment]
Hi,

regarding your first question:

It's possible to define your own filter method using function `makeFilter`.
Here is an example with a criterion from CORElearn:

```
reliefFequalK = makeFilter(
  name = "ReliefFequalK",
  desc = "ReliefF algorithm where k nearest instances have equal weight ",
  pkg  = "CORElearn",
  supported.tasks = c("classif", "regr"),
  supported.features = c("numerics", "factors"),
  fun = function(task, nselect, ...) {
    CORElearn::attrEval(getTaskFormula(task), data = getTaskData(task), estimator = "ReliefFequalK", ...)
  }
)

fv = generateFilterValuesData(iris.task, method = "ReliefFequalK")
fv
FilterValues:
Task: iris-example
          name    type ReliefFequalK
1 Sepal.Length numeric     0.3060741
2  Sepal.Width numeric     0.0857963
3 Petal.Length numeric     0.7590471
4  Petal.Width numeric     0.7369630
```

(I didn't check if `supported.tasks` and `supported.features` is set correctly, just copy-pasted from the interface to `FSelector::relief`.)

This file (https://github.com/mlr-org/mlr/blob/master/R/Filter.R) contains the code for all filter methods available through mlr, if you want to look at some more examples.

Would you maybe be interested in sending us a PR for the filter methods you are going to plug into mlr?

Cheers,
Julia

**Comment by andrewjohnlowe** [comment]
Hi Julia, thanks!

Can I use a similar recipe to fuse a custom filter method to a learner? If makeFilter creates a filter and registers it with (for example) the name "ReliefFequalK", can I fuse it with a learner in the same way as those from FSelector?

Any ideas on how to use bootstrapped filter measures in a filter fused with a learner? (2nd question) This would be very useful in cases (like mine) in which the filter measures are unstable.

What do you mean by a PR? Sorry, don't understand. Happy to help, though.

My main reason for preferring CORElearn over FSelector is that previously I've encountered problems that resulted in execution failures:
https://github.com/larskotthoff/fselector/issues/4
This isn't an issue with FSelector, but rather with RWeka. I haven't had time to follow-up with it, and instead found that CORElearn provided what I needed.

Best regards,

Andrew.

**Comment by schiffner** [comment]
Hi Andrew,

I'm very sorry for the delay. Meant to get back to you earlier but was a little ill this week.

> Can I use a similar recipe to fuse a custom filter method to a learner? If makeFilter creates a filter and registers it with (for example) the name "ReliefFequalK", can I fuse it with a learner in the same way as those from FSelector?

Yes. If you write your own filter via `makeFilter` you can use it like any other filter method already integrated in mlr. And particularly, you can fuse it with a learner by `makeFilterWrapper`.
For example:

```
lrn = makeFilterWrapper("classif.fnn", fw.method = "ReliefFequalK", fw.abs = 2)
m = train(lrn, iris.task)
m
Model for learner.id=classif.fnn.filtered; learner.class=FilterWrapper
Trained on: task.id = iris-example; obs = 150; features = 4
Hyperparameters: fw.method=ReliefFequalK,fw.abs=2
```

> Any ideas on how to use bootstrapped filter measures in a filter fused with a learner? (2nd question) This would be very useful in cases (like mine) in which the filter measures are unstable.

To be sure that I understood correctly. You want to do the following:
On the training set:
1) calculate bootstrapped feature importance
2) select the top features based on mean/median importance values
3) train the learner

On the test set:
4) predict on the same features used for learning

The FilterWrapper does not do bootstrapping itself. It simply applies the filter method before training the learner.

Maybe the easiest way for now would be to implement the bootstrapping within the function `fun` that you pass to `makeFilter`.

One issue with the FilterWrapper is that it currently "forgets" the importance values. So accessing the SD or IQRs of the feature importances is not totally easy at the moment.

> What do you mean by a PR? Sorry, don't understand. Happy to help, though.

Sorry, I meant a [pull request](https://help.github.com/articles/using-pull-requests/) or, more generally, if you would be willing to share your code.
My thinking was that if you were going to integrate some of the CORElearn filters for your purposes, we could all profit by making them a part of mlr.

Thanks,
Julia

**Comment by andrewjohnlowe** [comment]
Hi Julia,

Sorry for the delay in replying to your reply! I was on vacation, then had various deadlines to meet upon my return, so I've only had time to return to this now.

Upon reflection, probably the best way forward for me is to implement the method you described, that is, to implement the bootstrapping within the function fun that I pass to makeFilter. Alternatively, I can just live with the variability of these filter methods on my data, do repeated CV, and the uncertainty will just fold into my final error estimate.

I think you pretty much answered my question with the code snippets you've provided, so you can close this issue if you wish. Thanks for your help!

I'm writing a journal paper and intend to make the whole thing reproducible by publishing the R Markdown code, and I'd be happy if you want to make use of my code. I'm afraid I must admit that I'm not that familiar with git and GitHub for sharing code (although I intend to learn when I have the time!), so I don't know yet how to use these tools to contribute code. This XKCD cartoon describes where I am with git now: https://xkcd.com/1597/ !

Best regards,

Andrew.

**Comment by schiffner** [comment]
Hi Andrew,

thanks for your reply.

I will close this issue here.
I have opened another one regarding extraction of the filter scores from the FilterWrapper (#942).

> I'm writing a journal paper and intend to make the whole thing reproducible by publishing the R Markdown code, and I'd be happy if you want to make use of my code. I'm afraid I must admit that I'm not that familiar with git and GitHub for sharing code (although I intend to learn when I have the time!), so I don't know yet how to use these tools to contribute code. This XKCD cartoon describes where I am with git now: https://xkcd.com/1597/ !

:smile: I like the cartoon.
If you could drop us a note when your code is published, that would be great.

Thanks and best regards,
Julia
